File: sdk/python/harnessxml/parse.py
"""`.hxml` -> Harness.

Copyright 2026 VisML. SPDX-License-Identifier: Apache-2.0

Namespace-aware per §2.6: elements are matched on namespace URI plus local name,
never on prefix.

Per §2.8 an unrecognised element in the HarnessXML namespace is an ERROR
(HX-1003), not something to skip. That rule is enforced here, in the parser,
rather than left to the validator — a construct the parser silently dropped is
one the validator can never see.

Standard library only: xml.etree.ElementTree, with a line-number-tracking
parser so diagnostics can point at the offending element.
"""

from __future__ import annotations

import xml.etree.ElementTree as ET

from .diag import Diagnostics
from .model import (
    NS,
    Artifact,
    Cases,
    Edge,
    Harness,
    Loop,
    Metadata,
    Node,
    Port,
    Ref,
    Resource,
    Retry,
    Timeout,
    Wait,
)


def _annotate_lines(root: ET.Element, text: str) -> None:
    """Attach source line numbers to every element.

    ElementTree drops them, and subclassing XMLParser._start does not work
    because the C accelerator bypasses it — which is why an earlier version of
    this module reported every finding at line 0. A diagnostic without a
    location is a rule number, not a diagnostic (§13.5).

    So: run expat over the source to collect start-element lines in document
    order, then zip them against a depth-first preorder walk of the tree, which
    is the same order.
    """
    import xml.parsers.expat

    lines: list[int] = []
    parser = xml.parsers.expat.ParserCreate(namespace_separator="}")

    def start(_name, _attrs):
        lines.append(parser.CurrentLineNumber)

    parser.StartElementHandler = start
    try:
        parser.Parse(text.encode("utf-8"), True)
    except xml.parsers.expat.ExpatError:
        return  # already reported by ElementTree; leave lines at 0

    for el, line in zip(root.iter(), lines):
        el.set("__line__", str(line))


def _line(el: ET.Element) -> int:
    try:
        return int(el.get("__line__", "0"))
    except ValueError:
        return 0


def _local(tag: str) -> tuple[str | None, str]:
    """Split `{ns}local` into (ns, local)."""
    if tag.startswith("{"):
        ns, _, local = tag[1:].partition("}")
        return ns, local
    return None, tag


def _bool(el: ET.Element, key: str, default: bool) -> bool:
    v = el.get(key)
    if v == "true":
        return True
    if v == "false":
        return False
    return default


def _int(el: ET.Element, key: str) -> int | None:
    v = el.get(key)
    if v is None:
        return None
    try:
        return int(v)
    except ValueError:
        return None


def _float(el: ET.Element, key: str, default: float) -> float:
    v = el.get(key)
    if v is None:
        return default
    try:
        return float(v)
    except ValueError:
        return default


#: Elements whose only job is to contain others.
_CONTAINERS = {
    "metadata", "provenance", "tags", "security", "permission", "generator",
    "source", "signature", "resources", "artifacts", "nodes", "edges",
    "inputs", "outputs", "config", "description", "credential", "tag",
    "title", "author", "organization", "created", "modified", "license",
    "documentVersion",
}


def parse(text: str, diags: Diagnostics) -> Harness | None:
    """Parse a HarnessXML document. Returns None if it is not well-formed or
    has no `<harness>` root."""
    try:
        root = ET.fromstring(text)
    except ET.ParseError as e:
        line = e.position[0] if e.position else 1
        diags.error("HX-1001", line, f"not well-formed: {e}")
        return None
    _annotate_lines(root, text)

    ns, local = _local(root.tag)
    if local != "harness" or ns != NS:
        diags.error(
            "HX-1001",
            _line(root),
            f"root element is <{local}> in namespace {ns!r}; "
            f"expected <harness> in {NS!r}",
        )
        return None

    h = Harness(
        id=root.get("id", ""),
        spec_version=root.get("specVersion"),
        name=root.get("name"),
        entry=root.get("entry"),
    )

    for child in root:
        cns, cl = _local(child.tag)
        if cns != NS:
            diags.error(
                "HX-1006",
                _line(child),
                f"element <{cl}> is from a foreign namespace and is only "
                f"permitted inside <extension>",
            )
            continue
        if cl == "metadata":
            _metadata(child, h)
        elif cl == "security":
            pass
        elif cl == "resources":
            _resources(child, h, diags)
        elif cl == "artifacts":
            _artifacts(child, h, diags)
        elif cl == "nodes":
            _nodes(child, h, diags)
        elif cl == "edges":
            _edges(child, h, diags)
        elif cl == "extension":
            pass  # vendor content is not ours to interpret
        else:
            diags.error(
                "HX-1003",
                _line(child),
                f"unrecognised element <{cl}> inside <harness>; "
                f"an unknown construct must be rejected, never skipped",
            )
    return h


def _metadata(el: ET.Element, h: Harness) -> None:
    m = Metadata()
    for c in el:
        _, cl = _local(c.tag)
        text = (c.text or "").strip()
        if cl == "title":
            m.title = text
        elif cl == "description":
            m.description = text
        elif cl == "author":
            m.author = text
        elif cl == "organization":
            m.organization = text
        elif cl == "created":
            m.created = text
        elif cl == "modified":
            m.modified = text
        elif cl == "license":
            m.license = text
        elif cl == "documentVersion":
            m.document_version = text
        elif cl == "tags":
            m.tags = [(t.text or "").strip() for t in c]
    h.metadata = m


def _resources(el: ET.Element, h: Harness, diags: Diagnostics) -> None:
    for c in el:
        _, cl = _local(c.tag)
        if cl != "resource":
            diags.error("HX-1003", _line(c), f"unrecognised element <{cl}> inside <resources>")
            continue
        r = Resource(
            id=c.get("id", ""),
            type=c.get("type", ""),
            name=c.get("name"),
            provider=c.get("provider"),
            uri=c.get("uri"),
            line=_line(c),
        )
        for g in c:
            _, gl = _local(g.tag)
            if gl == "property":
                r.properties.append((g.get("name", ""), g.get("value", "")))
            elif gl == "credential":
                r.credential_ref = g.get("ref")
                r.credential_store = g.get("store")
        h.resources.append(r)


def _artifacts(el: ET.Element, h: Harness, diags: Diagnostics) -> None:
    for c in el:
        _, cl = _local(c.tag)
        if cl != "artifact":
            diags.error("HX-1003", _line(c), f"unrecognised element <{cl}> inside <artifacts>")
            continue
        a = Artifact(
            id=c.get("id", ""),
            type=c.get("type", ""),
            name=c.get("name"),
            uri=c.get("uri"),
            media_type=c.get("mediaType"),
            digest=c.get("digest"),
            classification=c.get("classification"),
            line=_line(c),
        )
        for g in c:
            _, gl = _local(g.tag)
            if gl == "property":
                a.properties.append((g.get("name", ""), g.get("value", "")))
        h.artifacts.append(a)


def _ports(el: ET.Element, want: str) -> list[Port]:
    out = []
    for c in el:
        _, cl = _local(c.tag)
        if cl != want:
            continue
        out.append(
            Port(
                name=c.get("name", ""),
                type=c.get("type"),
                required=_bool(c, "required", True),
                default=c.get("default"),
                value=c.get("value"),
                line=_line(c),
            )
        )
    return out


def _nodes(el: ET.Element, h: Harness, diags: Diagnostics) -> None:
    for c in el:
        _, cl = _local(c.tag)
        if cl != "node":
            diags.error("HX-1003", _line(c), f"unrecognised element <{cl}> inside <nodes>")
            continue
        n = Node(
            id=c.get("id", ""),
            type=c.get("type", ""),
            name=c.get("name"),
            impl=c.get("impl"),
            idempotent=_bool(c, "idempotent", True),
            join_policy=c.get("joinPolicy", "all"),
            quorum=_int(c, "quorum"),
            compensates=c.get("compensates"),
            line=_line(c),
        )
        for g in c:
            gns, gl = _local(g.tag)
            if gns != NS:
                diags.error("HX-1006", _line(g), f"foreign element <{gl}> outside <extension>")
                continue
            if gl == "description":
                n.description = (g.text or "").strip()
            elif gl == "inputs":
                n.inputs = _ports(g, "input")
            elif gl == "outputs":
                n.outputs = _ports(g, "output")
            elif gl == "config":
                n.config = [
                    (p.get("name", ""), p.get("value", ""))
                    for p in g
                    if _local(p.tag)[1] == "property"
                ]
            elif gl == "resourceRef":
                n.resource_refs.append(Ref(g.get("ref", ""), _line(g)))
            elif gl == "artifactRef":
                n.artifact_refs.append(Ref(g.get("ref", ""), _line(g)))
            elif gl == "guard":
                n.guard = g.get("when", "")
            elif gl == "retry":
                n.retry = Retry(
                    max_attempts=_int(g, "maxAttempts") or 1,
                    backoff=g.get("backoff", "exponential"),
                    initial_delay=g.get("initialDelay", "PT1S"),
                    max_delay=g.get("maxDelay"),
                    multiplier=_float(g, "multiplier", 2.0),
                    jitter=_bool(g, "jitter", True),
                    retry_on=(g.get("retryOn") or "").split(),
                )
            elif gl == "timeout":
                n.timeout = Timeout(
                    duration=g.get("duration", ""),
                    on_timeout=g.get("onTimeout", "fail"),
                )
            elif gl == "cases":
                cs = Cases(line=_line(g))
                for k in g:
                    _, kl = _local(k.tag)
                    if kl == "case":
                        cs.cases.append((k.get("when", ""), k.get("to", "")))
                    elif kl == "otherwise":
                        cs.otherwise = k.get("to")
                n.cases = cs
            elif gl == "loop":
                lp = Loop(
                    kind=g.get("kind", ""),
                    over=g.get("over"),
                    while_expr=g.get("while"),
                    count=_int(g, "count"),
                    max_iterations=_int(g, "maxIterations"),
                    var=g.get("var", "item"),
                    index_var=g.get("indexVar", "index"),
                    max_concurrency=_int(g, "maxConcurrency") or 1,
                    on_item_failure=g.get("onItemFailure", "fail"),
                    line=_line(g),
                )
                for k in g:
                    if _local(k.tag)[1] == "body":
                        lp.body = k.get("ref")
                n.loop = lp
            elif gl == "subworkflow":
                n.subworkflow_href = g.get("href")
            elif gl == "wait":
                n.wait = Wait(
                    duration=g.get("duration"),
                    until=g.get("until"),
                    event=g.get("event"),
                    line=_line(g),
                )
            elif gl == "security":
                pass
            elif gl == "extension":
                n.extensions.append(
                    (g.get("namespace", ""), g.get("required") == "true")
                )
            else:
                diags.error(
                    "HX-1003",
                    _line(g),
                    f"unrecognised element <{gl}> inside <node id='{n.id}'>",
                )
        h.nodes.append(n)


def _edges(el: ET.Element, h: Harness, diags: Diagnostics) -> None:
    for c in el:
        _, cl = _local(c.tag)
        if cl != "edge":
            diags.error("HX-1003", _line(c), f"unrecognised element <{cl}> inside <edges>")
            continue
        ty = c.get("type", "control")
        from .model import EDGE_TYPES

        if ty not in EDGE_TYPES:
            diags.error("HX-1003", _line(c), f"unrecognised edge type '{ty}'")
            continue
        h.edges.append(
            Edge(
                id=c.get("id"),
                from_=c.get("from", ""),
                to=c.get("to", ""),
                type=ty,
                from_port=c.get("fromPort"),
                to_port=c.get("toPort"),
                condition=c.get("condition"),
                line=_line(c),
            )
        )

File: sdk/go/parse.go
package harnessxml

// `.hxml` -> Harness.
//
// Copyright 2026 VisML. SPDX-License-Identifier: Apache-2.0
//
// Namespace-aware per §2.6: elements are matched on namespace URI plus local
// name, never on prefix.
//
// Per §2.8 an unrecognised element in the HarnessXML namespace is an ERROR
// (HX-1003), not something to skip. That rule is enforced here, in the parser,
// rather than left to the validator — a construct the parser silently dropped
// is one the validator can never see.
//
// encoding/xml gives no line numbers, so the parser tracks them itself from the
// decoder's byte offset. A diagnostic without a location is a rule number, not
// a diagnostic (§13.5).

import (
	"encoding/xml"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type parser struct {
	dec        *xml.Decoder
	diags      *Diagnostics
	lineStarts []int
}

func newLineIndex(src []byte) []int {
	starts := []int{0}
	for i, b := range src {
		if b == '\n' {
			starts = append(starts, i+1)
		}
	}
	return starts
}

func (p *parser) lineAt(offset int64) int {
	i := sort.SearchInts(p.lineStarts, int(offset))
	if i <= 0 {
		return 1
	}
	return i
}

func attr(se xml.StartElement, name string) string {
	for _, a := range se.Attr {
		if a.Name.Local == name && a.Name.Space != "xmlns" {
			return a.Value
		}
	}
	return ""
}

func attrPtr(se xml.StartElement, name string) *string {
	for _, a := range se.Attr {
		if a.Name.Local == name && a.Name.Space != "xmlns" {
			v := a.Value
			return &v
		}
	}
	return nil
}

func attrBool(se xml.StartElement, name string, def bool) bool {
	switch attr(se, name) {
	case "true":
		return true
	case "false":
		return false
	}
	return def
}

func attrInt(se xml.StartElement, name string) *int {
	v := attr(se, name)
	if v == "" {
		return nil
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return nil
	}
	return &n
}

func attrFloat(se xml.StartElement, name string, def float64) float64 {
	v := attr(se, name)
	if v == "" {
		return def
	}
	f, err := strconv.ParseFloat(v, 64)
	if err != nil {
		return def
	}
	return f
}

// Parse reads a HarnessXML document. It returns nil if the document is not
// well-formed or has no <harness> root; diagnostics always explain why.
func Parse(src []byte, diags *Diagnostics) *Harness {
	p := &parser{
		dec:        xml.NewDecoder(strings.NewReader(string(src))),
		diags:      diags,
		lineStarts: newLineIndex(src),
	}

	// Find the root.
	for {
		tok, err := p.dec.Token()
		if err != nil {
			diags.Errorf("HX-1001", 1, "not well-formed: %v", err)
			return nil
		}
		se, ok := tok.(xml.StartElement)
		if !ok {
			continue
		}
		line := p.lineAt(p.dec.InputOffset())
		if se.Name.Local != "harness" || se.Name.Space != NS {
			diags.Errorf("HX-1001", line,
				"root element is <%s> in namespace %q; expected <harness> in %q",
				se.Name.Local, se.Name.Space, NS)
			return nil
		}
		return p.harness(se, line)
	}
}

func (p *parser) harness(se xml.StartElement, line int) *Harness {
	h := &Harness{
		ID:          attr(se, "id"),
		SpecVersion: attrPtr(se, "specVersion"),
		Name:        attr(se, "name"),
		Entry:       attr(se, "entry"),
	}

	for {
		tok, err := p.dec.Token()
		if err != nil {
			if err.Error() == "EOF" {
				break
			}
			p.diags.Errorf("HX-1001", p.lineAt(p.dec.InputOffset()), "not well-formed: %v", err)
			return h
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "harness" {
				return h
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Space != NS {
				p.diags.Errorf("HX-1006", l,
					"element <%s> is from a foreign namespace and is only permitted inside <extension>",
					t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			switch t.Name.Local {
			case "metadata":
				p.metadata(h)
			case "security", "extension":
				_ = p.dec.Skip()
			case "resources":
				p.resources(h)
			case "artifacts":
				p.artifacts(h)
			case "nodes":
				p.nodes(h)
			case "edges":
				p.edges(h)
			default:
				p.diags.Errorf("HX-1003", l,
					"unrecognised element <%s> inside <harness>; an unknown construct must be rejected, never skipped",
					t.Name.Local)
				_ = p.dec.Skip()
			}
		}
	}
	return h
}

func (p *parser) text() string {
	var b strings.Builder
	depth := 0
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return strings.TrimSpace(b.String())
		}
		switch t := tok.(type) {
		case xml.CharData:
			if depth == 0 {
				b.Write(t)
			}
		case xml.StartElement:
			depth++
		case xml.EndElement:
			if depth == 0 {
				return strings.TrimSpace(b.String())
			}
			depth--
		}
	}
}

func (p *parser) metadata(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "metadata" {
				return
			}
		case xml.StartElement:
			switch t.Name.Local {
			case "title":
				h.Metadata.Title = p.text()
			case "description":
				h.Metadata.Description = p.text()
			case "author":
				h.Metadata.Author = p.text()
			case "organization":
				h.Metadata.Organization = p.text()
			case "created":
				h.Metadata.Created = p.text()
			case "modified":
				h.Metadata.Modified = p.text()
			case "license":
				h.Metadata.License = p.text()
			case "documentVersion":
				h.Metadata.DocumentVersion = p.text()
			case "tags":
				p.tags(h)
			default:
				_ = p.dec.Skip()
			}
		}
	}
}

func (p *parser) tags(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "tags" {
				return
			}
		case xml.StartElement:
			if t.Name.Local == "tag" {
				h.Metadata.Tags = append(h.Metadata.Tags, p.text())
			} else {
				_ = p.dec.Skip()
			}
		}
	}
}

func (p *parser) resources(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "resources" {
				return
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Local != "resource" {
				p.diags.Errorf("HX-1003", l, "unrecognised element <%s> inside <resources>", t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			r := Resource{
				ID: attr(t, "id"), Type: attr(t, "type"), Name: attr(t, "name"),
				Provider: attr(t, "provider"), URI: attr(t, "uri"), Line: l,
			}
			p.resourceBody(&r)
			h.Resources = append(h.Resources, r)
		}
	}
}

func (p *parser) resourceBody(r *Resource) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "resource" {
				return
			}
		case xml.StartElement:
			switch t.Name.Local {
			case "property":
				r.Properties = append(r.Properties, Property{attr(t, "name"), attr(t, "value")})
				_ = p.dec.Skip()
			case "credential":
				r.CredentialRef = attr(t, "ref")
				r.CredentialStore = attr(t, "store")
				_ = p.dec.Skip()
			default:
				_ = p.dec.Skip()
			}
		}
	}
}

func (p *parser) artifacts(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "artifacts" {
				return
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Local != "artifact" {
				p.diags.Errorf("HX-1003", l, "unrecognised element <%s> inside <artifacts>", t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			a := Artifact{
				ID: attr(t, "id"), Type: attr(t, "type"), Name: attr(t, "name"),
				URI: attr(t, "uri"), MediaType: attr(t, "mediaType"),
				Digest: attr(t, "digest"), Classification: attr(t, "classification"),
				Line: l,
			}
			p.artifactBody(&a)
			h.Artifacts = append(h.Artifacts, a)
		}
	}
}

func (p *parser) artifactBody(a *Artifact) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "artifact" {
				return
			}
		case xml.StartElement:
			if t.Name.Local == "property" {
				a.Properties = append(a.Properties, Property{attr(t, "name"), attr(t, "value")})
			}
			_ = p.dec.Skip()
		}
	}
}

func (p *parser) nodes(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "nodes" {
				return
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Local != "node" {
				p.diags.Errorf("HX-1003", l, "unrecognised element <%s> inside <nodes>", t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			n := Node{
				ID: attr(t, "id"), Type: attr(t, "type"), Name: attr(t, "name"),
				Impl: attr(t, "impl"), Idempotent: attrBool(t, "idempotent", true),
				JoinPolicy: "all", Quorum: attrInt(t, "quorum"),
				Compensates: attr(t, "compensates"), Line: l,
			}
			if jp := attr(t, "joinPolicy"); jp != "" {
				n.JoinPolicy = jp
			}
			p.nodeBody(&n)
			h.Nodes = append(h.Nodes, n)
		}
	}
}

func (p *parser) nodeBody(n *Node) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "node" {
				return
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Space != NS {
				p.diags.Errorf("HX-1006", l, "foreign element <%s> outside <extension>", t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			switch t.Name.Local {
			case "description":
				n.Description = p.text()
			case "inputs":
				n.Inputs = p.ports("inputs", "input")
			case "outputs":
				n.Outputs = p.ports("outputs", "output")
			case "config":
				n.Config = p.properties("config")
			case "resourceRef":
				n.ResourceRefs = append(n.ResourceRefs, Ref{attr(t, "ref"), l})
				_ = p.dec.Skip()
			case "artifactRef":
				n.ArtifactRefs = append(n.ArtifactRefs, Ref{attr(t, "ref"), l})
				_ = p.dec.Skip()
			case "guard":
				g := attr(t, "when")
				n.Guard = &g
				_ = p.dec.Skip()
			case "retry":
				ma := 1
				if v := attrInt(t, "maxAttempts"); v != nil {
					ma = *v
				}
				backoff := "exponential"
				if b := attr(t, "backoff"); b != "" {
					backoff = b
				}
				initial := "PT1S"
				if d := attr(t, "initialDelay"); d != "" {
					initial = d
				}
				n.Retry = &Retry{
					MaxAttempts: ma, Backoff: backoff, InitialDelay: initial,
					MaxDelay: attr(t, "maxDelay"), Multiplier: attrFloat(t, "multiplier", 2),
					Jitter:  attrBool(t, "jitter", true),
					RetryOn: strings.Fields(attr(t, "retryOn")),
				}
				_ = p.dec.Skip()
			case "timeout":
				ot := "fail"
				if v := attr(t, "onTimeout"); v != "" {
					ot = v
				}
				n.Timeout = &Timeout{Duration: attr(t, "duration"), OnTimeout: ot}
				_ = p.dec.Skip()
			case "cases":
				n.Cases = p.cases(l)
			case "loop":
				n.Loop = p.loop(t, l)
			case "subworkflow":
				n.Subworkflow = attr(t, "href")
				_ = p.dec.Skip()
			case "wait":
				n.Wait = &Wait{
					Duration: attr(t, "duration"), Until: attr(t, "until"),
					Event: attr(t, "event"), Line: l,
				}
				_ = p.dec.Skip()
			case "security":
				_ = p.dec.Skip()
			case "extension":
				n.Extensions = append(n.Extensions, Extension{
					Namespace: attr(t, "namespace"),
					Required:  attr(t, "required") == "true",
				})
				_ = p.dec.Skip()
			default:
				p.diags.Errorf("HX-1003", l,
					"unrecognised element <%s> inside <node id='%s'>", t.Name.Local, n.ID)
				_ = p.dec.Skip()
			}
		}
	}
}

func (p *parser) ports(container, want string) []Port {
	var out []Port
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return out
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == container {
				return out
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Local == want {
				out = append(out, Port{
					Name: attr(t, "name"), Type: attr(t, "type"),
					Required: attrBool(t, "required", true),
					Default:  attrPtr(t, "default"), Value: attrPtr(t, "value"),
					Line: l,
				})
			}
			_ = p.dec.Skip()
		}
	}
}

func (p *parser) properties(container string) []Property {
	var out []Property
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return out
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == container {
				return out
			}
		case xml.StartElement:
			if t.Name.Local == "property" {
				out = append(out, Property{attr(t, "name"), attr(t, "value")})
			}
			_ = p.dec.Skip()
		}
	}
}

func (p *parser) cases(line int) *Cases {
	c := &Cases{Line: line}
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return c
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "cases" {
				return c
			}
		case xml.StartElement:
			switch t.Name.Local {
			case "case":
				c.Cases = append(c.Cases, Case{attr(t, "when"), attr(t, "to")})
			case "otherwise":
				o := attr(t, "to")
				c.Otherwise = &o
			}
			_ = p.dec.Skip()
		}
	}
}

func (p *parser) loop(se xml.StartElement, line int) *Loop {
	l := &Loop{
		Kind: attr(se, "kind"), Over: attr(se, "over"), While: attr(se, "while"),
		Count: attrInt(se, "count"), MaxIterations: attrInt(se, "maxIterations"),
		Var: "item", IndexVar: "index", MaxConcurrency: 1, OnItemFailure: "fail",
		Line: line,
	}
	if v := attr(se, "var"); v != "" {
		l.Var = v
	}
	if v := attr(se, "indexVar"); v != "" {
		l.IndexVar = v
	}
	if v := attrInt(se, "maxConcurrency"); v != nil {
		l.MaxConcurrency = *v
	}
	if v := attr(se, "onItemFailure"); v != "" {
		l.OnItemFailure = v
	}
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return l
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "loop" {
				return l
			}
		case xml.StartElement:
			if t.Name.Local == "body" {
				l.Body = attr(t, "ref")
			}
			_ = p.dec.Skip()
		}
	}
}

func (p *parser) edges(h *Harness) {
	for {
		tok, err := p.dec.Token()
		if err != nil {
			return
		}
		switch t := tok.(type) {
		case xml.EndElement:
			if t.Name.Local == "edges" {
				return
			}
		case xml.StartElement:
			l := p.lineAt(p.dec.InputOffset())
			if t.Name.Local != "edge" {
				p.diags.Errorf("HX-1003", l, "unrecognised element <%s> inside <edges>", t.Name.Local)
				_ = p.dec.Skip()
				continue
			}
			ty := attr(t, "type")
			if ty == "" {
				ty = "control"
			}
			known := false
			for _, k := range EdgeTypes {
				if k == ty {
					known = true
				}
			}
			if !known {
				p.diags.Errorf("HX-1003", l, "unrecognised edge type '%s'", ty)
				_ = p.dec.Skip()
				continue
			}
			h.Edges = append(h.Edges, Edge{
				ID: attr(t, "id"), From: attr(t, "from"), To: attr(t, "to"), Type: ty,
				FromPort: attr(t, "fromPort"), ToPort: attr(t, "toPort"),
				Condition: attr(t, "condition"), Line: l,
			})
			_ = p.dec.Skip()
		}
	}
}

var _ = fmt.Sprintf
